src/txrequest.cpp

```diff
@@ -0,0 +1,593 @@
+        std::transform(selected.begin(), selected.end(), std::back_inserter(ret), [](const Announcement* ann) {
+            return GenTxid{ann->m_is_wtxid, ann->m_txhash};
+        });
+        return ret;
+    }
+
+    void RequestedTx(NodeId peer, const uint256& txhash, std::chrono::microseconds expiry)
+    {
+        auto it = m_index.get<ByPeer>().find(ByPeerView{peer, true, txhash});
+        if (it == m_index.get<ByPeer>().end()) {
+            // There is no CANDIDATE_BEST announcement, look for a _READY or _DELAYED instead. If the caller only
+            // ever invokes RequestedTx with the values returned by GetRequestable, and no other non-const functions
+            // other than ForgetTxHash and GetRequestable in between, this branch will never execute (as txhashes
+            // returned by GetRequestable always correspond to CANDIDATE_BEST announcements).
+
+            it = m_index.get<ByPeer>().find(ByPeerView{peer, false, txhash});
+            if (it == m_index.get<ByPeer>().end() || (it->m_state != State::CANDIDATE_DELAYED &&
+                it->m_state != State::CANDIDATE_READY)) {
+                // There is no CANDIDATE announcement tracked for this peer, so we have nothing to do. Either this
+                // txhash wasn't tracked at all (and the caller should have called ReceivedInv), or it was already
+                // requested and/or completed for other reasons and this is just a superfluous RequestedTx call.
+                return;
+            }
+
+            // Look for an existing CANDIDATE_BEST or REQUESTED.
+            auto it_old = m_index.get<ByTxHash>().lower_bound(ByTxHashView{txhash, State::CANDIDATE_BEST, 0});
+            if (it_old != m_index.get<ByTxHash>().end() && it_old->m_txhash == txhash) {
+                if (it_old->m_state == State::CANDIDATE_BEST) {
+                    // The data structure's invariants require that there can be at most one CANDIDATE_BEST or one
+                    // REQUESTED announcement per txhash (but not both simultaneously), so we have to convert any
+                    // existing CANDIDATE_BEST to another CANDIDATE_* when constructing another REQUESTED.
+                    // It doesn't matter whether we pick CANDIDATE_READY or _DELAYED here, as SetTimePoint()
+                    // will correct it at GetRequestable() time. If time only goes forward, it will always be
+                    // _READY, so pick that to avoid extra work in SetTimePoint().
+                    Modify<ByTxHash>(it_old, [](Announcement& ann) { ann.m_state = State::CANDIDATE_READY; });
+                } else if (it_old->m_state == State::REQUESTED) {
+                    // As we're no longer waiting for a response to the previous REQUESTED announcement, convert it
+                    // to COMPLETED. This also helps guaranteeing progress.
+                    Modify<ByTxHash>(it_old, [](Announcement& ann) { ann.m_state = State::COMPLETED; });
```

Comment:
I guess this ideally should happen atomically with setting `it` to REQUESTED? Otherwise, some call in-between these two actions may, for example, make another request for the same tx? (as they see there is no currently `REQUESTED`)

Reply:
`TxRequestTracker` is thread-agnostic so that's not allowed. The caller needs external synchronization if multiple threads are going to be accessing it.

Reply:
I see.